pkg/promtail/targets/file/tailer.go

```diff
@@ -105,12 +98,21 @@ func (t *tailer) run() {
 		case <-positionWait.C:
 			err := t.markPositionAndSize()
 			if err != nil {
-				level.Error(t.logger).Log("msg", "error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				level.Error(t.logger).Log("msg", "position timer: error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				// To prevent a deadlock on stopping the tailer we need to launch a thread to consume any unread lines
+				go func() {
+					for range t.tail.Lines {}
+				}()
```

Comment:
I'm finding it a little hard to follow the logic here, unfortunately. I think this could benefit from some comments around the lifecycle of this loop. I guess the way it works now is that one of the two select cases will eventually fail/be closed? 

Also, even though you only need the goroutine to drain the lines channel in this specific branch, IMO it'd be more resilient to future changes if it was drained in a defer to the whole run loop.